remove.go

```diff
@@ -0,0 +1,76 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+
+	"gopkg.in/yaml.v2"
+)
+
+func removeOldVersions(runner *Environ, workspace string, vargs GAE) error {
+	// read in the app.yaml file to grab the module/service mame we just deployed
+	appLoc := filepath.Join(workspace, vargs.Dir, "app.yaml")
+	appFile, err := os.Open(appLoc)
+	if err != nil {
+		return fmt.Errorf("error: %s\n", err)
+	}
+	defer appFile.Close()
+	var appStruct struct {
+		Service string `yaml:"service"`
+		Module  string `yaml:"module"`
+	}
+	err = yaml.NewDecoder(appFile).Decode(&appStruct)
+	if err != nil {
+		return fmt.Errorf("error: %s\n", err)
+	}
+
+	service := appStruct.Service
+	if service == "" {
+		service = appStruct.Module
+	}
+
+	// look  up existing versions for given service ordered by create time desc
+	var versionJSON bytes.Buffer
+	sout := runner.stdout
+	runner.stdout = &versionJSON
+	err = runner.Run(vargs.GCloudCmd, "app", "versions", "list",
+		"--service", service, "--project", vargs.Project,
+		"--format", "json", "--sort-by", "~version.createTime", "--quiet")
+	if err != nil {
+		return fmt.Errorf("error: %s\n", err)
+	}
+
+	var results []struct {
+		ID           string  `json:"id"`
+		TrafficSplit float64 `json:"traffic_split"`
+	}
+	err = json.NewDecoder(&versionJSON).Decode(&results)
+	if err != nil {
+		return err
+	}
+
+	var toDelete []string
+	for i, res := range results {
+		// keep newer versions, the newly deployed version or anything that has traffic
+		if i < vargs.MaxVersions || res.ID == vargs.Version || res.TrafficSplit > 0 {
+			continue
```

Comment:
If one of the versions intended to be deleted still has traffic, does it mean that we'll increment `i` but not delete that version leaving us with less versions deleted than suppose to?

Reply:
i also stared at this for a minute but it seems okay to me. If you have MaxVersions = 10, you'd be left with 8 idle versions, 1 new version, and the current 1 serving traffic.

Reply:
This was one compromise to deal with the case that one of the "older" versions that should be flagged for deletion actually is (1) somehow the version we just deployed or (2) is currently serving _any_ traffic.

I'll make a note in the `MaxVersions` docs to say that any older versions that are serving traffic will not be deleted.

Reply:
I'm adding:

> If any of the "older" versions that should be deleted is actually
    // serving traffic, they will not be deleted. This may result in the actual version
    // count being higher than the max listed here.